File: July2021/CustomSortString.java
class Solution {
  // O(Len(Str + Order))
  public String customSortString(String order, String str) {
    int[] freq = new int[26];

    // TC : O(len of Str)
    for(char c : str.toCharArray()){
      freq[c-'a']++;
    }

    StringBuilder ans = new StringBuilder();

    // TC : O(Len(orderStr + Str))
    for(char orderChar: order.toCharArray()) {

      while(freq[orderChar-'a']>0) {
        ans.append(orderChar);
        freq[orderChar-'a']--;
      }
    }

    // TC : O(Len(Str))
    for(int i=0;i<26;i++){
      int freqC = freq[i];
      while(freqC>0) {
        ans.append((char)(i+'a'));
        freqC--;
      }
    }
    return ans.toString();
  }
}

File: July2021/C++/CustomSortString.cpp
class Solution
{
public:
    string customSortString(string order, string str)
    {
        string ans = "";
        map<char, int> m;
        for (auto x : str)
        {
            m[x]++;
        }
        for (auto x : order)
        {
            if (m.find(x) != m.end())
            {
                auto temp = m.find(x);
                int count = temp->second;
                ans += string(count, x);
                m.erase(x);
            }
        }
        for (auto x : m)
        {
            ans += string(x.second, x.first);
        }
        return ans;
    }
};